pkg/service/manifest/response.go

```diff
@@ -98,7 +146,107 @@ func (s Service) buildCredentialResponse(ctx context.Context, applicantDID, mani
 }
 
-func buildDenialCredentialResponse(manifestID, applicationID, reason string, failedOutputDescriptorIDs ...string) (*manifest.CredentialResponse, error) {
+func applyIssuanceTemplate(
+	credentialRequest *credential.CreateCredentialRequest,
+	template *issuing.IssuanceTemplate,
+	templateMap map[string]*issuing.CredentialTemplate,
+	od manifest.OutputDescriptor,
+	applicationJSON map[string]any,
+	credManifest manifest.CredentialManifest,
+	application manifest.CredentialApplication,
+) error {
+	credentialRequest.Issuer = template.Issuer
+
+	ct, ok := templateMap[od.ID]
+	if !ok {
+		logrus.Warnf("Did not find output_descriptor with ID \"%s\" in template. Skipping application.", od.ID)
+		return nil
+	}
+	c, err := getCredential(applicationJSON, ct, credManifest, application.PresentationSubmission)
+	if err != nil {
+		return err
+	}
+	for k, v := range ct.Data.Claims.Data {
+		claimValue := v
+		vs, ok := v.(string)
+		if ok {
+			if strings.HasPrefix(vs, "$") {
+				claimValue, err = jsonpath.JsonPathLookup(c, vs)
+				if err != nil {
+					return errors.Wrapf(err, "looking up json path \"%s\" for key=\"%s\"", vs, k)
+				}
+			}
+		}
+		credentialRequest.Data[k] = claimValue
+	}
+
+	if ct.Expiry.Time != nil {
+		credentialRequest.Expiry = ct.Expiry.Time.Format(time.RFC3339)
+	}
+
+	if ct.Expiry.Duration != nil {
+		credentialRequest.Expiry = time.Now().Add(*ct.Expiry.Duration).Format(time.RFC3339)
+	}
+
+	credentialRequest.Revocable = ct.Revocable
+	return nil
+}
+
+func getCredential(
+	applicationJSON map[string]any,
+	ct *issuing.CredentialTemplate,
+	credManifest manifest.CredentialManifest,
+	submission *exchange.PresentationSubmission,
+) (any, error) {
+	if ct.Data.CredentialInputDescriptor != "" {
+		if credManifest.PresentationDefinition == nil {
+			return nil, errors.New("cannot provide input descriptor when manifest does not have presentation definition")
+		}
+
+		// Lookup the claim that's sent in the submission.
+		for _, descriptor := range submission.DescriptorMap {
+			if descriptor.ID == ct.Data.CredentialInputDescriptor {
+				c, err := jsonpath.JsonPathLookup(applicationJSON, descriptor.Path)
+				if err != nil {
+					return nil, errors.Wrapf(
+						err,
+						"looking up json path \"%s\" for submission=\"%s\"",
+						descriptor.Path,
+						descriptor.ID,
+					)
+				}
+				claim, err := fromFormat(exchange.JWTFormat(descriptor.Format), c)
+				if err != nil {
+					return nil, err
+				}
+				return claim, nil
+			}
+		}
+		return nil, errors.Errorf(
+			"could not find credential for input_descriptor=\"%s\"",
+			ct.Data.CredentialInputDescriptor,
+		)
+	}
+	return nil, nil
+}
+
+func fromFormat(format exchange.JWTFormat, claim any) (any, error) {
+	switch format {
+	case exchange.JWTVC:
+		_, token, err := util.ParseJWT(keyaccess.JWT(claim.(string)))
+		if err != nil {
+			return nil, errors.Wrapf(err, "parsing jwt as %s", exchange.JWTVC)
+		}
+		return token.PrivateClaims()["vc"], nil
```

Comment:
might want to break this into two lines, there's a chance the `["vc"]` lookup can crash